node/src/main/scala/coop/rchain/node/NodeRuntime.scala

```diff
@@ -701,7 +699,7 @@ object NodeRuntime {
         implicit val ra = rpConfAsk
         implicit val eb = eventPublisher
         implicit val sc = synchronyConstraintChecker
-        CasperLaunch[F](casperInit)
+        new DefaultCasperLaunch[F](conf.casper).launch()
```

Comment:
`F` is still abstract at this point - why committing to a concrete implementation at this point?

Reply:
WDYM? We're at the edge of the world, hence the concrete instance.

Reply:
You're not actually at the edge of the world which is in `main` - you're in `setupNodeProgramF` which is abstract in `F` . So IMO the cleaner version would be dependent on algebra you created eg `def setupNodeProgramF[F: CasperLaunch] = ...` or, explicitly `def setupNodeProgramF[F[_]](launcher: CasperLauch[F]) = ...` if you treat it more as an utitlity. This concern is also mirrored in @sebtomba review (and if you coded it this way, it'd solve both problems)

Reply:
IMO there is another valid approach here - if we actually wanted to treat `NodeRuntime` as the end of the known world then `DefaultCasperLaunch` could be ditched in favor of `implicit val casperLaunchInstance: CasperLaunch[TaskEnv] = ...` . I am not sure which one is better tbh

Reply:
I agree. We should treat `CasperLaunch` as any other dependency.

Reply:
setupNodeProgramF is the producer of casper launch, not the consumer. I've pulled the actual `.launch()` call to the top level (to `main`) to make the launch more explicit. This also makes the trait used.

If you guys mean it should say 

```
CasperLaunch[F]
```

instead of

```
new DefaultCasperLaunch(casperConf)
```

then I beg to differ: we should not use the `apply` method on a TF's companion object for anything else than the implicit summoner (`def apply[F[_](implicit ev: Foo[F]) = ev`), or we risk ad-lib instantiation of TF algebra instances down in the program. This, it would seem, is in opposition to what we currently have in some places, e.g.:

https://github.com/rchain/rchain/blob/f802f1f4bacd2b7e2492f7cfa6a8eb2e562bcb48/node/src/main/scala/coop/rchain/node/NodeRuntime.scala#L711

Reply:
>  we should not use the apply method on a TF's companion object for anything else than the implicit summoner (def apply[F[_](implicit ev: Foo[F]) = ev), or we risk ad-lib instantiation of TF

I wasn't proposing that - I agree with you here. What lost me here is that in the previous version `setupNodeProgramF` was actually creating **launch** effect (which made a lot of sense given the name, barring the fact that it was hidden under `apply`). In your design what we have is effectively a lawless algebra wich advertises that it can be interpreted under any `F` to produce `launch` side-effect - whatever this means is left unspecified (and I agree, semantics here are hard to capture - just like, say, context shift). So my previous comment about passing it as a bound is, in fact, wrong because there is no `setup` needed. Now, this method is not in the strict sense of the word a producer of algebra. It is just its, hmm, by-passer :-) So how about:

```scala
object CasperLaunch {
  def of[F: WhatYouGot: AndMore](conf: CaperConf) = //the real deal
}
// default is a bit misleading as well because there is no "alternative"
// if you wanted to eg. turn it off in you could simply do
// new CasperLaunch[F] { def launch() = Applicative[F].unit }
// but it's not like you would be offering this as an alternative to the default
```

If what I said doesn't make sense then just ignore me :-) overall, the PR is fine

Reply:
If you insist, I can alias the `new DefaultCasperLaunch()` constructor as `CasperLaunc.of`, but that would mean duplicating the bounds list. Used as a convention, that would mean we need to declare (and change when needed) each TF bounds list twice.

Maybe a better suffix/prefix than `Default*` would do? `*Impl` is out of question for me :) FWIW, in general, if there's just one choice I consider it a good default choice :D

> setupNodeProgramF was actually creating launch effect

it was executing the very same launch effect under the covers of innocuous `CasperLaunch[F]` :)

> algebra wich advertises that it can be interpreted under any F

This is true of every TF trait :) Hopefully, at least, or else the TF flavor at hand is not composable (can't stub without providing the deps; deps propagate, need to express them twice (in trait and impl)).

> lawless algebra

We're not yet at the stage to have any lawful algebras, and most TF-'algebras' aren't. Should we be calling them TF-traits? Or are we fine with 'lawless' being implied by 'algebra' used near 'TF'? (I'm fine with both :)) 

> overall, the PR is fine

May I have a :heavy_check_mark: ? :)

Reply:
> If you insist, I can alias the new DefaultCasperLaunch() constructor as CasperLaunc.of, but that would mean duplicating the bounds list

I was thinking about anonymous class - no duplication needed and no name needs to be invented :-) Anyways, if you think `Default` is ok then I believe in your judgement.

> or else the TF flavor at hand is not composable

Here we differ a bit. Constraints on derived operations do not hurt composability. Contravariance does :-)
```scala
trait Lauch[F[_]] { self =>
  def launch(finalizer: F[Unit]): F[LaunchResult]

  final def mapk[G[_]](fk: F ~> G): Alg1[G] = new Alg1[G] {
    override def launch(finalizer: G[Unit]): G[LaunchResult] = sys.error("impossible")
  }
}
```

In this case it's fine to be lawless (I can't see any useful law attached to an algebra with a single operation returning `F[Unit]` :-)), just as it is fine for a `ContextShift` to be lawless. I am not sure about the "most aren't" part, but I am fine with "algebra" terminology :-).

My dubious nitpicking aside - a change for the better!